[ThinLTO][Bitcode] Generate import type in bitcode

mingmingl-llvm opened the pull request:
For distributed ThinLTO, the LTO indexing step generates combined summary for each module, and postlink pipeline reads the combined summary which stores the information for link-time optimization.

This patch populates the 'import type' of a summary in bitcode, and updates bitcode reader to parse the bit correctly.

mingmingl-llvm committed:
[ThinLTO]Record import type (declaration or definition) in GlobalValueSummary::GVFlags

- This change doesn't set the bit in the summaries, just make the code changes.

mingmingl-llvm committed:
[ThinLTO]Generate import type in bitcode writer

mingmingl-llvm committed:
function import changes

mingmingl-llvm committed:
'git merge main' and resolve merge conflict

mingmingl-llvm committed:
Merge branch 'main' into users/minglotus-6/spr/summary1

mingmingl-llvm committed:
rename Dec to Decl

mingmingl-llvm committed:
Merge branch 'main' into users/minglotus-6/spr/summary1

mingmingl-llvm committed:
Merge branch 'users/minglotus-6/spr/summary1' into users/minglotus-6/spr/summary2

mingmingl-llvm committed:
simplify code

mingmingl-llvm committed:
Flag gate the import-declaration change.

mingmingl-llvm committed:
stage changes

mingmingl-llvm committed:
Add test coverage for import delcaration

mingmingl-llvm committed:
add test coverage for function alias

mingmingl-llvm committed:
llvm-lto test coverage

mingmingl-llvm committed:
add test coverage for debugging log, and a RUN line for internalization

mingmingl-llvm committed:
Merge branch 'main' into users/minglotus-6/spr/summary1

minglotus-6 commented:
Tagging @jvoung for review since I cannot add you, thanks!

minglotus-6 commented on llvm/include/llvm/IR/ModuleSummaryIndex.h:
```diff
@@ -1272,6 +1272,12 @@ using ModulePathStringTableTy = StringMap<ModuleHash>;
+/// A set of global value summary pointers.
+using GVSummaryPtrSet = SmallPtrSet<GlobalValueSummary *, 4>;
+
+/// The key is module path, and value is a set of global value summary pointers.
```
Good question. I think it makes sense to use a set of pointers as suggested, given `ModuleToGVSummaryPtrSet` is computed using pointers as opposed to GUIDs. Going to make the change.

While it's possible to have {mod_A, GUID_foo} and {mod_B, GUID_foo} entries because of comdats, each `GlobalValueSummary*` will be associated with exactly one module.

minglotus-6 replied:
Done. Going to update the subsequent patch after updating this PR.

jvoung replied:
Ah got it, thanks for the clarification!

mingmingl-llvm committed:
resolve review feedback

minglotus-6 commented on llvm/lib/Bitcode/Writer/BitcodeWriter.cpp:
```diff
@@ -428,6 +428,10 @@ class IndexBitcodeWriter : public BitcodeWriterBase {
   /// The combined index to write to bitcode.
   const ModuleSummaryIndex &Index;
 
+  /// For each module, provides the set of global value summaries for which the
```
Updated the comment to "When writing combined summaries, provides the set of global value summaries...".

While we write a subset of the index only for combined summaries, 'combined summary' is slightly preferred as it hints this map doesn't apply for per-module summary or the index.

mingmingl-llvm reviewed (commented):
thanks! ptal.

mingmingl-llvm committed:
Merge branch 'users/minglotus-6/spr/summary1' into users/minglotus-6/spr/summary2

mingmingl-llvm committed:
Diffbase is updated (https://github.com/llvm/llvm-project/pull/87600/commits/2c38b47e4526686a2353a1cdd282352a2927f055).
Update this one accordingly

minglotus-6 commented:
thanks for the reviews @jvoung ! I'll hold this patch and plan to land this together with the follow-up one (https://github.com/llvm/llvm-project/pull/88024), primarily because regression tests are added there; meanwhile this gives time for more feedbacks :)

teresajohnson reviewed (commented):
This change looks fine but see my comment on PR88024 for a suggestion on testing for this one (moving a couple things from that patch to this one and submitting this one second).

mingmingl-llvm committed:
resolve review feedback

mingmingl-llvm committed:
update this patch as the second one

mingmingl-llvm committed:
update stale comment and use 'DAG' for check lines

jvoung commented on llvm/include/llvm/Transforms/IPO/FunctionImport.h:
```diff
@@ -216,7 +216,8 @@ void gatherImportedSummariesForModule(
     const FunctionImporter::ImportMapTy &ImportList,
-    std::map<std::string, GVSummaryMapTy> &ModuleToSummariesForIndex);
+    std::map<std::string, GVSummaryMapTy> &ModuleToSummariesForIndex,
+    GVSummaryPtrSet &DecSummaries);
```
Can you add a documentation for the DecSummaries parameter here too? (similar to the ThinLTOCodeGenerator.h comment, might be useful to indicate it's a subset of ModuleToSummariesForIndex).

minglotus-6 replied:
done.

mingmingl-llvm committed:
add comment for 'DecSummaries' parameter in FunctionImport.h

jvoung commented on llvm/lib/LTO/LTO.cpp:
```diff
@@ -1399,18 +1399,21 @@ class lto::ThinBackendProc {
     if (EC)
       return errorCodeToError(EC);
 
     // TODO: Serialize declaration bits to bitcode.
```
Can the TODO be removed?

jvoung commented on llvm/test/ThinLTO/X86/import_callee_declaration.ll:
```diff
@@ -48,20 +48,20 @@
-; MAIN-DIS-NOT: [[LARGEFUNC:\^[0-9]+]] = gv: (guid: 2418497564662708935, summaries: (function: (module: [[LIBMOD]], flags: ({{.*}} importType: declaration), insts: 8, {{.*}})))
-; MAIN-DIS-NOT: [[LARGEINDIRECT:\^[0-9]+]] = gv: (guid: 14343440786664691134, summaries: (function: (module: [[LIBMOD]], flags: ({{.*}} importType: declaration), insts: 8, {{.*}})))
-; MAIN-DIS-NOT: [[LARGEINDIRECTALIAS:\^[0-9]+]] = gv: (guid: 16730173943625350469, summaries: (alias: (module: [[LIBMOD]], flags: ({{.*}} importType: declaration)
+; MAIN-DIS: [[LARGEFUNC:\^[0-9]+]] = gv: (guid: 2418497564662708935, summaries: (function: (module: [[LIBMOD]], flags: ({{.*}} importType: declaration), insts: 8, {{.*}})))
```
It might be interesting to test happens if "calleeAddrs" had just an alias of a function and not the original function (right now both large_indirect_callee and large_indirect_callee_alias are there?), but could be okay to test that later too.

minglotus-6 replied:
> As shown by the updated test, function alias summary has 'null' aliasee. I'm preparing a fix.

Update on this: it's valid for an aliasee to be 'null'. https://github.com/llvm/llvm-project/blob/3e15c97fa3812993bdc319827a5c6d867b765ae8/llvm/test/ThinLTO/X86/distributed_indexes.ll#L45-L49 is a test case

jvoung reviewed (commented):
Nice!

mingmingl-llvm committed:
run 'git merge main'

mingmingl-llvm committed:
resolve review feedback

mingmingl-llvm committed:
Update regression test:
1. Set 'import-instr-evolution-factor' to 1.0 (default 0.7) to make sure
  import limit remains 7 (otherwise import limit decays through each
  edge)
2. Updated bar* function names
3. Explain why aliasee is null for bar* functions
